File: neetcode/java/String/ValidParentheses.java
// https://leetcode.com/problems/valid-parentheses
// https://neetcode.io/problems/validate-parentheses?list=neetcode150

import java.util.Scanner;
import java.util.Stack;

public class ValidParentheses {
  public static void main(String[] args) {
    String str = getInput();
    boolean result = validateParentheses(str);
    System.out.println(result);
  }

  private static boolean validateParentheses(String str) {
    Stack<Character> stack = new Stack<>();

    for (char c : str.toCharArray()) {
      if (isOpenParentheses(c)) {
        stack.push(c);
      } else if (isCloseParentheses(c)) {
        if (stack.isEmpty() || stack.pop() != complementOf(c)) {
          return false;
        }
      }
    }

    return stack.isEmpty();
  }

  private static char complementOf(char c) {
    switch (c) {
      case ')': return '(';
      case '}': return '{';
      case ']': return '[';
      default: return '\0';
    }
  }

  private static boolean isOpenParentheses(char c) {
    return c == '(' || c == '{' || c == '[';
  }

  private static boolean isCloseParentheses(char c) {
    return c == ')' || c == '}' || c == ']';
  }

  private static String getInput() {
    Scanner sc = new Scanner(System.in);
    System.out.print("Enter a string with parentheses: ");
    String str = sc.nextLine();
    sc.close();

    return str;
  }
}

File: neetcode/python/String/valid_parentheses.py
# https://leetcode.com/problems/valid-parentheses
# https://neetcode.io/problems/validate-parentheses?list=neetcode150

class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False

        bracket_map = {
          ')': '(',
          '}': '{',
          ']': '['
        }
        stack = []

        for char in s:
            if char in bracket_map.values():  # opening bracket
                stack.append(char)
            elif char in bracket_map:         # closing bracket
                if not stack or stack.pop() != bracket_map[char]:
                    return False

        return not stack
